### ISSUE TITLE
Encountered bad file descriptor error

### ISSUE THREAD

**Issue by ikenchina** [facebook/rocksdb#7735]
> Note: Please use Issues only for bug reports. For questions, discussions, feature requests, etc. post to dev group: https://groups.google.com/forum/#!forum/rocksdb or https://www.facebook.com/groups/rocksdb.dev

Rocksdb returned IO error when wrote a batch. 
`
IO error: While pread offset 19862407 len 1146: /db/db/1/1/1958633.sst: Bad file descriptor
`

Rocksdb switched to read-only mode, and has not invoked `rocksdb::EventListener::OnBackgroundError`
It works fine after reopen it.

**Comment by ajkr** [comment]
Do you use `Merge()` with `max_successive_merges` or in-place memtable updates? If so it is possible reads are done in the write path. If the read encounters an error, we fail the write and enter read-only mode (as you observed).

**Comment by ikenchina** [comment]
I didn't use `Merge()` and in-place memtable, i only write a new key for testing performence.

**Comment by ajkr** [comment]
Oh, then it is an error encountered during a background operation. Then the error is saved, the DB enters read-only mode , and future write operations are returned the saved error.

**Comment by ikenchina** [comment]
So, does rocksdb invoke `rocksdb::EventListener::OnBackgroundError` for this issue?

**Comment by ajkr** [comment]
> So, does rocksdb invoke rocksdb::EventListener::OnBackgroundError for this issue?

Yes, it is supposed to call `OnBsckgroundError()` in this scenario. I just tried to test it on master branch by removing some files while the DB was running and verified it got called. What version are you using?

> And can i ignore it as the code below

Probably, but it's not really recommended. That capability (manipulating background error from a callback) was added for outdated political reasons and should be removed. An alternative is setting `paranoid_checks = false` to avoid going into read-only mode. But it's not entirely clear why a bad file descriptor would become readable if we simply proceed without any mitigation.

I think closing/re-opening the DB is a good choice. That will cause RocksDB to get all new file descriptors, providing a better chance of fixing things. It also avoids relying on any unsafe/rarely used code paths.
